src/nvim/shada.c

```diff
@@ -4117,3 +4053,225 @@ static bool shada_removable(const char *name)
   xfree(new_name);
   return retval;
 }
+
+/// Initialize ShaDa jumplist entries.
+///
+/// @param[in,out]  jumps           Array of ShaDa entries to set.
+/// @param[in]      removable_bufs  Cache of buffers ignored due to their
+///                                 location.
+///
+/// @return number of jumplist entries
+static inline size_t shada_init_jumps(
+    PossiblyFreedShadaEntry *jumps, khash_t(bufset) *const removable_bufs)
```

Comment:
Would be nice to avoid the need to pass around `remove_bufs`. It's a bit weird that it's needed everywhere. Also, without it, we could avoid lifting `khash_t(bufset)`  out of the shada module.

remove_bufs is documented as "Cache of buffers ignored". Why is it so important to cache it? Just make the shada module figure it out internally. It's ok if `FOR_ALL_BUFFERS` needs to happen multiple times, that's not really a bottleneck and not worth the burden of passing around this awkward data structure. Not to mention, cache management itself is risky and should be avoided unless absolutely necessary.